### PR TITLE
Sign multipart objects in the gate

### LINKED ISSUE #975: Sign multiparts and do not use sessions
## Is your feature request related to a problem? Please describe.
S3 gateway slices big objects if it is required (multipart API is used). It does almost all the work neofs-node usually does, so no code simplification from opening a session is achieved, so why not just finish the job that is 90% done?  

## Describe the solution you'd like
Sign multiparts before sending to NeoFS, do not open/store/cache a session, it will not be required.

## Describe alternatives you've considered
Do nothing, it already works (but looks strange when you tell about it works to _any_ neofs dev).

## Additional context
Can be a solution for https://github.com/nspcc-dev/neofs-s3-gw/issues/969 (however, neofs-node is also guilty here).

### PR DESCRIPTION
**Pull request by smallhive**
Closes #975.

### CODE REVIEW

File: internal/neofs/neofs.go [hunk 1]
```diff
@@ -310,6 +342,10 @@ func (x *NeoFS) CreateObject(ctx context.Context, prm layer.PrmObjectCreate) (oi
 			prm.Payload = bytes.NewReader(obj.Payload())
 			obj.SetPayloadSize(uint64(len(obj.Payload())))
 
+			if err := x.signMultipartObject(&obj, x.signer(ctx)); err != nil {
+				return oid.ID{}, errors.New("object sign failed")
```
Comment (roman-khimov) [inline internal/neofs/neofs.go:346]: failed to sign object

File: internal/neofs/neofs.go [hunk 2]
```diff
@@ -246,6 +247,37 @@ func (x *NeoFS) DeleteContainer(ctx context.Context, id cid.ID, token *session.C
...
 }
 
+func (x *NeoFS) signMultipartObject(header *object.Object, signer neofscrypto.Signer) error {
+	var (
+		payloadChecksum = sha256.New()
+		sig             neofscrypto.Signature
+	)
+
+	payloadChecksum.Write(header.Payload())
+	header.SetPayloadChecksum(checksum.NewFromHash(checksum.SHA256, payloadChecksum))
```
Comment (roman-khimov) [inline internal/neofs/neofs.go:257]: I thought you have them somewhere already for multiparts. Or you can calculate them on the go, just like regular ones. The rationale is that you'd have warmer caches then.

File: internal/neofs/neofs.go (hunk 1 above)
Comment (smallhive) [reply internal/neofs/neofs.go]: Updated

File: internal/neofs/neofs.go (hunk 2 above)
Comment (smallhive) [reply internal/neofs/neofs.go]: It exists, but it calculates on the fly during payload writing to the network. Updated to recalculated hash

Comment (roman-khimov) [reply internal/neofs/neofs.go]: But tz is still the same.

Comment (smallhive) [reply internal/neofs/neofs.go]: Yes, we calculate `tz` for the whole Big object. Separately for each part, we didn't calculate it before. The node did it on my behalf. Now we calculate it in place

File: api/layer/object.go [hunk 3]
```diff
@@ -488,17 +488,32 @@ func (n *layer) objectPutAndHash(ctx context.Context, prm PrmObjectCreate, bktIn
...
+	n.prepareAuthParameters(ctx, &prm.PrmAuth, bktInfo.Owner)
+
+	prm.Payload = wrapReader(prm.Payload, 64*1024, func(buf []byte) {
 		if prm.Multipart != nil {
 			for _, h := range prm.Multipart.MultipartHashes {
 				h.Write(buf)
 			}
 		}
 	})
+
```
Comment (carpawell) [inline api/layer/object.go:511]: space change

File: internal/neofs/neofs.go [hunk 4]
```diff
@@ -246,6 +247,35 @@ func (x *NeoFS) DeleteContainer(ctx context.Context, id cid.ID, token *session.C
...
+	var (
+		sig neofscrypto.Signature
+	)
+
+	header.SetPayloadChecksum(checksum.NewFromHash(checksum.SHA256, payloadHash))
+
+	if x.IsHomomorphicHashingEnabled() {
+		var homoChecksum = tz.New()
+
+		homoChecksum.Write(header.Payload())
```
Comment (carpawell) [inline internal/neofs/neofs.go:260]: `header` but `Payload()`?

File: internal/neofs/neofs.go [hunk 5]
```diff
@@ -246,6 +247,35 @@ func (x *NeoFS) DeleteContainer(ctx context.Context, id cid.ID, token *session.C
...
+	if err != nil {
+		return fmt.Errorf("calculate ID: %w", err)
+	}
+	header.SetID(id)
+
+	if err = sig.Calculate(signer, id.Marshal()); err != nil {
+		return fmt.Errorf("sign object ID: %w", err)
+	}
+
+	header.SetSignature(&sig)
```
Comment (carpawell) [inline internal/neofs/neofs.go:274]: just `CalculateAndSetID` and not 4 method calls above?

**Review by carpawell — COMMENTED** [review]
Dont we need some tests for it? Test code can be unified with SDK/node. Otherwise, it is impossible to keep everything in sync to me. Multiple codes do the same thing.

File: api/layer/object.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (smallhive) [reply api/layer/object.go:511]: It is a completely new function! But git shows it strangely. BTW removed an empty row

File: internal/neofs/neofs.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (smallhive) [reply internal/neofs/neofs.go:260]: Correct, is it wrong? Each part should have a tz checksum for it. For instance, the slicer works in the same way

**Comment by smallhive** [comment]
> Dont we need some tests for it? Test code can be unified with SDK/node. Otherwise, it is impossible to keep everything in sync to me. Multiple codes do the same thing.

I checked the code with s3 gate tests locally and GH pipeline did it as well

File: api/layer/object.go (hunk 3 above)
Comment (smallhive) [reply api/layer/object.go:511]: Removed

File: internal/neofs/neofs.go (hunk 4 above)
Comment (smallhive) [reply internal/neofs/neofs.go]: But variable name, replaced with `obj`

File: internal/neofs/neofs.go (hunk 5 above)
Comment (smallhive) [reply internal/neofs/neofs.go]: It seems shorter, replaced

File: api/layer/object.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply api/layer/object.go]: diff only cares about lines, so if there is any similar code already the result can be surprising.

File: internal/neofs/neofs.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply internal/neofs/neofs.go]: Sure. But you've just moved SHA256 calculation out of this function. Why not doing this for tz?

File: internal/neofs/neofs.go (hunk 5 above)
Comment (carpawell) [reply internal/neofs/neofs.go]: @smallhive, my bad, i meant `SetIDWithSignature` but copied a different method name. it allows even fewer code lines for the same functionality

**Comment by carpawell** [comment]
> s3 gate tests locally and GH pipeline

@smallhive, i guess without https://github.com/nspcc-dev/neofs-s3-gw/pull/974 it also worked well but contained errors, right? I do not mean this must be done in this PR (at least something should appear in SDK) but I think we can discuss it at least. @roman-khimov

File: internal/neofs/neofs.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (smallhive) [reply internal/neofs/neofs.go]: It was a double-content calculation in the first version. I separated it into another function to remove it. This version contains less code. Replaced

File: internal/neofs/neofs.go (hunk 5 above)
Comment (smallhive) [reply internal/neofs/neofs.go]: This is even better, updated. We should make such changes in the Slicer as well. I inspired it's code

**Comment by smallhive** [comment]
> @smallhive, i guess without #974 it also worked well but contained errors, right? I do not mean this must be done in this PR (at least something should appear in SDK) but I think we can discuss it at least. @roman-khimov

It's make sence

File: internal/neofs/neofs.go [hunk 6]
```diff
@@ -306,13 +320,31 @@ func (x *NeoFS) CreateObject(ctx context.Context, prm layer.PrmObjectCreate) (oi
 		}
 
 		if prm.Multipart.Link != nil {
+			// pl := obj.Payload()
```
Comment (carpawell) [inline internal/neofs/neofs.go:323]: not needed?

Comment (smallhive) [reply internal/neofs/neofs.go]: indeed

**Review by roman-khimov — APPROVED** [review]

**Review by carpawell — APPROVED** [review]

**Pull request merged**